text/3621-derive-smart-pointer.md

```diff
@@ -0,0 +1,676 @@
+stabilize `StableDeref`. This way, there's no stable way to use
+`#[derive(SmartPointer)]` types with `Pin` coercions.
+
+This solutions isn't a problem for the Linux Kernel right now because our custom
+`Arc` happens to be implicitly pinned for convenience reasons, but if it wasn't,
+then we would need coercions of `Pin`-wrapped values.
+
+## Negative trait bounds?
+
+There are also various solutions that involve negative trait bounds. For
+example, you might instead modify `CoerceUnsized` like this:
+```rust
+// Permit going from `Pin<impl Unpin>` to` Pin<impl Unpin>`
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: Deref<Target: Unpin>,
+    U: Deref<Target: Unpin>,
+{ }
+
+// Permit going from `Pin<impl !Unpin>` to `Pin<impl !Unpin>`
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: core::ops::Deref<Target: !Unpin>,
+    U: core::ops::Deref<Target: !Unpin>,
+{ }
+```
+This way, you can only coerce pinned pointers when this doesn't change whether
+the target type is `Unpin`.
+
+It would solve the unsoundness, but it does have the disadvantage of having no
+path to making these pinned coercions possible for smart pointers that don't
+have malicious `Deref` implementations. Another downside of this solution is
+that it's a breaking change, because it disallows these pinned coercions even
+with the standard library pointers, which allows them today.
+
+There are also other variations on the negative trait bounds, which become
+different implementations of the "Don't allow Pin coercions with custom smart
+pointers" solution.
+
+This solution is discussed in more details in [the pre-RFC for stabilizing the
+underlying traits][pre-rfc].
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+One of the design goals of this RFC is that it should make this feature
+available to crates without significantly limiting how the underlying traits
+can evolve. The authors hope that we will find a way to stabilize the
+underlying traits in the future.
+
+One of the things that is left out of scope of this RFC is coercions involving
+custom transparent containers similar to [`Cell`]. They require an
+implementation of [`CoerceUnsized`] without [`DispatchFromDyn`].
+
+There is a reasonable change that we may be able to lift some of [the
+restrictions][input-requirements] on the shape of the struct as well. The
+current restrictions are just whatever [`DispatchFromDyn`] requires today, and
+proposals for relaxing them have been seen before (e.g., in the
+[pre-RFC][pre-rfc].)
+
+One example of a restriction that we could lift is the restrction that there is
+only one non-zero-sized field. This could allow implementations of `Rc` and
+`Arc` that store the value and refcount in two different allocations, like how
+the C++ `shared_ptr` works.
+```rust
+#[derive(SmartPointer)]
+pub struct Rc<T: ?Sized> {
+    refcount: NonNull<Refcount>,
+    value: NonNull<T>,
+}
+```
```

Comment:
i don't think DispatchFromDyn prevented this? after all `std::rc::Rc` is allowed to take an allocator and `Rc` is shaped like this:

```rust
pub struct Rc<T: ?Sized, A: Allocator = Global> {
    ptr: NonNull<RcBox<T>>,
    phantom: PhantomData<RcBox<T>>,
    alloc: A,  // <-- can have any size and alignment
}
```

Reply:
The `Rc` type only implements `DispatchFromDyn` when the allocator is zero-sized.
```Rust
#[unstable(feature = "coerce_unsized", issue = "18598")]
impl<T: ?Sized + Unsize<U>, U: ?Sized, A: Allocator> CoerceUnsized<Rc<U, A>> for Rc<T, A> {}

#[unstable(feature = "dispatch_from_dyn", issue = "none")]
impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Rc<U>> for Rc<T> {}
```

Reply:
okay thanks for the correction.